core/vm/evm.go

```diff
@@ -503,7 +503,65 @@ func (evm *EVM) Create(caller ContractRef, code []byte, gas uint64, value *big.I
 		return nil, common.ZeroAddr, 0, err
 	}
 	contractAddr = crypto.CreateAddress(caller.Address(), evm.StateDB.GetNonce(internalAddr), code)
-	return evm.create(caller, &codeAndHash{code: code}, gas, value, contractAddr)
+	if _, err := contractAddr.InternalAddress(); err == nil {
+		return evm.create(caller, &codeAndHash{code: code}, gas, value, contractAddr)
+	}
+
+	// Calculate the gas required for the keccak256 computation of the input data.
+	gasCost, err := calculateKeccakGas(code)
+	if err != nil {
+		return nil, common.ZeroAddr, 0, err
+	}
+
+	// attempt to grind the address
+	grindData, contractAddr, remainingGas, err := evm.attemptGrindContractCreation(caller, internalAddr, gas, gasCost, code)
+	if err != nil {
+		return nil, common.ZeroAddr, 0, err
+	}
+
+	gas = remainingGas
+
+	return evm.create(caller, &codeAndHash{code: grindData}, gas, value, contractAddr)
+
+}
+
+// calculateKeccakGas calculates the gas required for performing a keccak256 hash on the given data.
+// It returns the total gas cost and any error that may occur during the calculation.
+func calculateKeccakGas(data []byte) (int64, error) {
+	// Base gas for keccak256 computation.
+	keccakBaseGas := int64(params.Sha3Gas)
+	// Calculate the number of words (rounded up) in the data for gas calculation.
+	wordCount := (len(data) + 31) / 32 // Round up to the nearest word
+	return keccakBaseGas + int64(wordCount)*int64(params.Sha3WordGas), nil
+}
+
+// attemptContractCreation tries to create a contract address by iterating through possible nonce values.
+// It returns the modified data for contract creation and any error encountered.
+func (evm *EVM) attemptGrindContractCreation(caller ContractRef, internalAddr common.InternalAddress, gas uint64, gasCost int64, code []byte) ([]byte, common.Address, uint64, error) {
+	senderAddress := caller.Address()
+	nonce := evm.StateDB.GetNonce(internalAddr)
+	// Iterate through possible nonce values to find a suitable contract address.
+	for i := 0; i < params.MaxAddressGrindAttempts; i++ {
+		// Append the iteration count to the data, altering it for each attempt.
+		grindData := append(code, byte(i))
+		// Generate a potential contract address.
+		contractAddr := crypto.CreateAddress(senderAddress, nonce, grindData)
```

Comment:
What if we use crypto.CreateAddress2 here, with `i` as the salt? I think the result would be the same, but we wouldn't be modifying any contract code

Reply:
Refactored to use create2

Reply:
Not sure compute adress for create2 is not the same at all…. Needs testing

Reply:
Agreed, should be thorough on this